rfcs/0017-state-snapshotting.md

```diff
@@ -0,0 +1,244 @@
+out-of-scope for now. Partial snapshotting can be done first and in parallel
+with implementation of transition frontier persistance to make the full-offline
+version much quicker to implement afterwards.
+
+## Motivation
+
+[motivation]: #motivation
+
+As stated above, snaphotting parts of our application state will give us a
+mechanism for more quickly reproducing edge cases and errors as they arive.
+
+There are several stages. Each of which provides a checkpoint that gives us
+useful features.
+
+[partial]: #partial
+### Partial Snapshotting:
+
+Snapshot enough state to be able to produce the immediate next transition, and
+then halt.
+
+Why is this useful? We've wasted a lot of time waiting for the network to reach
+a certain failing block. This feature would save us time in the future. We can
+also easily make unit tests around edge-cases we fix by saving snapshots of
+those cases.
+
+[full-offline]: #full-offline
+### Full-offline Snapshotting
+
+Snapshot enough state to recreate a full-node that can continue from the point
+of the snapshot to produce more than just one more transition.
+
+Why is this useful? Sometimes we need more context than the exact block that
+fails. This would give us that context.
+
+[full-online]: #full-online
+### Full-online Snapshotting
+
+Snapshot enough state to recreate a full-node that can continue from the point
+of the snapshot and can reconnect to a network of nodes.
+
+This task would require synchronizing state between nodes in a network and
+would require a significant amount of effort until we finish implementing
+genesis. This is out-of-scope for the near future, and out-of-scope in this RFC.
+
+In the near term, we should complete both partial and full-offline snapshotting.
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+For each component, we need to:
+
+a. Properly marshall/unmarshall/copy each piece of data.
+
+b. Put the node into what is potentially a special state using the unmarshalled
+snapshot data instead of the normal happy path of the application.
+
+### Partial Snapshotting:
+
+We can make a record with all the information we need to store called
+`Snapshot.t`. The record will contain all the pieces of state called out below
+in some form or another which will be determined in implementation or
+described below.
+
+The following is what is believed to be an exhaustive list of the pieces of
+state needed for partial snapshotting support:
+
+1. Breadcrumb (protocol state + staged ledger)
+2. Root snarked ledger (the database)
+3. Consensus local state
+4. The current time
+5. `Coda_main`'s `Config.t`
+6. Proposal Data
+7. Transaction pool
+8. Snark pool
+
+When we want to start the node from the snapshot, we need to boot it into a
+special state that will execute the proposer to generate the next state and
+transition and then stall afterwards.
+
+For marshalling and unmarshalling and copying the data:
+
+* `bin_io` can be trivally used for (1), (3), (6)
```

Comment:
Don't think you would want to bin_io a breadcrumb. Seems expensive. We should implement the persistent transition frontier to implement the serialization of breadcrumbs.